### ISSUE TITLE
raw ingest failure

### ISSUE THREAD

**Issue by bkmartinjr** [single-cell-data/TileDB-SingleCell#69]
Ingesting a test case failed - perhaps due to the dense .raw.X?

```
In [4]: tiledbsc.SOMA("/tmp/test1").from_anndata(ad)
START  SOMA.from_ann
  START  DECATEGORICALIZING
  FINISH DECATEGORICALIZING TIME 0.063
  START  WRITING /tmp/test1
    START  WRITING /tmp/test1/X/data from <class 'numpy.ndarray'>
    FINISH WRITING /tmp/test1/X/data TIME 36.410
    START  WRITING /tmp/test1/obs
    FINISH WRITING /tmp/test1/obs TIME 0.633
    START  WRITING /tmp/test1/var
    FINISH WRITING /tmp/test1/var TIME 0.057
    START  WRITING /tmp/test1/obsm/X_umap
    Annotation matrix obsm/X_umap has shape (6288, 2)
    FINISH WRITING /tmp/test1/obsm/X_umap TIME 0.039
    START  WRITING /tmp/test1/raw
    START  WRITING /tmp/test1/raw/X/data from <class 'numpy.ndarray'>
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Input In [4], in <cell line: 1>()
----> 1 tiledbsc.SOMA("/tmp/test1").from_anndata(ad)

File ~/TileDB-SingleCell/apis/python/src/tiledbsc/soma.py:86, in SOMA.from_anndata(self, anndata)
     83     print(f"START  SOMA.from_ann")
     85 anndata = self.decategoricalize(anndata)
---> 86 self.write_tiledb_group(anndata)
     88 if self.verbose:
     89     print(util.format_elapsed(s,"FINISH  SOMA.from_ann"))

File ~/TileDB-SingleCell/apis/python/src/tiledbsc/soma.py:320, in SOMA.write_tiledb_group(self, anndata)
    318 if anndata.raw != None:
    319     raw_group_uri = self.uri + '/raw'
--> 320     self.write_raw_group(raw_group_uri, anndata.raw)
    321     soma_group.add(uri=raw_group_uri, relative=False, name="raw")
    323 # ----------------------------------------------------------------

File ~/TileDB-SingleCell/apis/python/src/tiledbsc/soma.py:457, in SOMA.write_raw_group(self, raw_group_uri, raw)
    454 raw_group = tiledb.Group(raw_group_uri, mode="w", ctx=self.ctx)
    456 X_uri = os.path.join(raw_group_uri, "X")
--> 457 self.write_X_group(X_uri, raw.X, raw.obs_names, raw.var_names)
    458 raw_group.add(uri=X_uri, relative=False, name="X")
    460 var_uri = os.path.join(raw_group_uri, 'var')

File ~/TileDB-SingleCell/apis/python/src/tiledbsc/soma.py:347, in SOMA.write_X_group(self, X_group_uri, X, obs_index, var_index)
    345 if X is not None:
    346     X_data_uri = X_group_uri + "/data"
--> 347     self.write_X_array(X_data_uri, X, obs_index, var_index)
    348     X_group.add(uri=X_data_uri, relative=False, name="data")
    350 X_group.close()

File ~/TileDB-SingleCell/apis/python/src/tiledbsc/soma.py:372, in SOMA.write_X_array(self, X_array_uri, X, obs_names, var_names)
    370     self.__ingest_coo_data_string_dims_rows_chunked(X_array_uri, X, obs_names, var_names)
    371 else:
--> 372     self.__ingest_coo_data_string_dims_whole(X_array_uri, X, obs_names, var_names)
    374 if self.verbose:
    375     print(util.format_elapsed(s, f"    FINISH WRITING {X_array_uri}"))

File ~/TileDB-SingleCell/apis/python/src/tiledbsc/soma.py:860, in SOMA.__ingest_coo_data_string_dims_whole(self, uri, mat, row_names, col_names)
    850 """
    851 Convert ndarray/(csr|csc)matrix to coo_matrix and ingest into TileDB.
    852 
   (...)
    856 :param col_names: List of column names.
    857 """
    859 assert len(row_names) == mat.shape[0]
--> 860 assert len(col_names) == mat.shape[1]
    862 mat_coo = scipy.sparse.coo_matrix(mat)
    863 d0 = row_names[mat_coo.row]

AssertionError:  

In [5]: ad.raw.X
Out[5]: 
array([[  0.,   0., 198., ...,   0.,   0.,   0.],
       [  0.,   0., 175., ...,   0.,   0.,   0.],
       [  0.,   0., 227., ...,   0.,   0.,   0.],
       ...,
       [  0.,   0., 192., ...,   0.,   0.,   0.],
       [  0.,   0.,  33., ...,   0.,   0.,   0.],
       [  0.,   0., 231., ...,   0.,   0.,   0.]], dtype=float32)
```

Dataset provided to @johnkerl offline

**Comment by johnkerl** [comment]
The odd thing is that we handle dense `X` ok — e.g. `pbmc-small.h5ad` and `pbmc3k_processed.h5ad` have dense `X` (although not dense `raw/X`) — and it’s the same method (just different arguments) for `X` and `raw/X` — so the density of `raw/X` isn’t a problem per se

The issue (surprisingly) involves `decategoricalize`:

```
>>> import tiledbsc as t
>>> import anndata as ad

>>> a = ad.read_h5ad('/Users/johnkerl/scdata/anndata-aux/dense-x.h5ad')
>>> s = t.SOMA('foo')

>>> b = s.decategoricalize(a)
  START  DECATEGORICALIZING
  FINISH DECATEGORICALIZING TIME 0.124

>>> len(a.raw.obs_names)
6288
>>> len(a.raw.var_names)
45768

>>> len(b.raw.obs_names)
6288
>>> len(b.raw.var_names)
0
```

**Comment by johnkerl** [comment]
When we decategoricalize `raw`
https://github.com/single-cell-data/TileDB-SingleCell/blob/main/apis/python/src/tiledbsc/soma.py#L195
somehow the `raw.var_names` is getting clobbered. There are no other arguments to the `anndata.Raw` constructor we can use, and the `raw.var_names` is immutable post-construct:

```
>>> b.raw.var_names = a.raw.var_names
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: can't set attribute
```

which is frustrating.

-- However -- the decategoricalization code predates https://github.com/single-cell-data/TileDB-SingleCell/pull/66 and while we do still need `decategoricalize`, perhaps it can be broken up and spread out to various array-to-tiledb callsites that need it ... 👀

**Comment by johnkerl** [comment]
TL;DR:

* We have a "blanket" method to convert non-tiledb-able things like categorical strings to Python strings, etc, "all at once" via `anndata = decategoricalize(anndata)`
* This iterates over all the `obs`, `varm`, `raw`, etc etc etc
* Inside that function we have `newraw = ad.Raw(...)` and it appears that the `ad.Raw` constructor somehow 'forgets' the `raw.var_names` :( 
* But a possible solution is that all the conversion of things like categorical strings to Python strings doesn't need to be done 'all at once', but rather, can be done as each component array is ingested.

**Comment by johnkerl** [comment]
https://github.com/single-cell-data/TileDB-SingleCell/pull/70
